xray/grpc.go

```diff
@@ -0,0 +1,174 @@
+	"net/http/httptrace"
+	"net/url"
+	"strconv"
+	"strings"
+
+	"github.com/golang/protobuf/proto"
+	"go.uber.org/multierr"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/aws/aws-xray-sdk-go/header"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+)
+
+// UnaryClientInterceptor provides gRPC unary client interceptor.
+func UnaryClientInterceptor(host string) grpc.UnaryClientInterceptor {
+	return func(ctx context.Context, method string, req, reply interface{}, cc *grpc.ClientConn, invoker grpc.UnaryInvoker, opts ...grpc.CallOption) error {
+		return Capture(ctx, host, func(ctx context.Context) error {
+			seg := GetSegment(ctx)
+			if seg == nil {
+				return errors.New("failed to record gRPC transaction: segment cannot be found")
+			}
+
+			ct, e := NewClientTrace(ctx)
+			if e != nil {
+				return e
+			}
+			md := metadata.Pairs(TraceIDHeaderKey, seg.DownstreamHeader().String())
+			ctx2 := metadata.NewOutgoingContext(httptrace.WithClientTrace(ctx, ct.httpTrace), md)
+
+			seg.Lock()
+			seg.Namespace = "remote"
+			seg.GetHTTP().GetRequest().URL = "grpc://" + host + method
+			seg.GetHTTP().GetRequest().Method = method
+			seg.Unlock()
+
+			err := invoker(ctx2, method, req, reply, cc, opts...)
+
+			recordContentLength(seg, reply)
+			if err != nil {
+				classifyErrorStatus(seg, err)
+				ct.subsegments.GotConn(nil, err)
+			}
+
+			return err
+		})
+	}
+}
+
+func unaryServerInterceptorWithConfig(sn SegmentNamer, cfg *Config) grpc.UnaryServerInterceptor {
+	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (resp interface{}, err error) {
+		md, ok := metadata.FromIncomingContext(ctx)
+
+		var traceID string
+		if ok && len(md.Get(TraceIDHeaderKey)) == 1 {
+			traceID = md.Get(TraceIDHeaderKey)[0]
+		}
+		traceHeader := header.FromString(traceID)
+
+		var host string
+
+		if len(md.Get(":authority")) == 1 {
+			host = md.Get(":authority")[0]
+		}
+		requestURL := url.URL{
+			Scheme: "grpc",
+			Host:   host,
+			Path:   info.FullMethod,
+		}
+		name := sn.Name(host)
+
+		ctx = context.WithValue(ctx, RecorderContextKey{}, cfg)
+
+		var seg *Segment
+		ctx, seg = NewSegmentFromHeader(ctx, name, &http.Request{
+			Host:   host,
+			URL:    &requestURL,
+			Method: info.FullMethod,
+		}, traceHeader)
+		defer seg.Close(nil)
+
+		seg.Lock()
+		seg.GetHTTP().GetRequest().ClientIP, seg.GetHTTP().GetRequest().XForwardedFor = clientIPFromGrpcMetadata(md)
+		seg.GetHTTP().GetRequest().URL = requestURL.String()
+		seg.GetHTTP().GetRequest().Method = info.FullMethod
+		if len(md.Get("user-agent")) == 1 {
+			seg.GetHTTP().GetRequest().UserAgent = md.Get("user-agent")[0]
+		}
+		seg.Unlock()
+
+		resp, err = handler(ctx, req)
+		if err != nil {
+			classifyErrorStatus(seg, err)
+		}
+		recordContentLength(seg, resp)
+		if headerErr := addResponseTraceHeader(ctx, seg, traceHeader); headerErr != nil {
+			err = multierr.Combine(err, headerErr)
```

Comment:
Can we avoid adding this new library dependency on `multierr`?

Reply:
As described here https://github.com/aws/aws-xray-sdk-go/pull/292#discussion_r623120247
At this line there're two possible error sources:
1. The error returned by the service handler itself
1. The error returned by calling `grpc.SendHeader`

I'm not sure how to report multiple errors within an interceptor. Maybe change this to print a line of error message when failling to send headers? What do you think?

Reply:
I think we can drop the error, it seems pretty rare. Alternatively a debug log message could be ok too. But yeah we shouldn't add it to the `err` here. It's not quite as important if `err` is actually set but imagine if its `nil` - then it means our interceptor changes the status of the actual gRPC request to an error which will probably cause problems in user code. The interceptor should only record issues but not introduce errors to the business logic like that.

Reply:
Okay, using the internal logger to print the debug message.